HW0/P5/P5.txt

```diff
@@ -1 +1,22 @@
 # Your answers here.
+1) 
+It will take 255 seconds.
+
+2)
+Yes, it will be faster. Now there swill be 7+1=8 cashiers in total. Let each cashier takes 256/8=32 bags and add them up which will take 32-1=31 seconds. Now we have 8 numbers. Split them into pairs and get the own sum of each pair. Repeat the last step until we get the final sum. This will take log(8)/log(2) = 3 seconds. Thus, the total time is 31+3 = 34 seconds.
+
+3)
+It will take log(256)/log(2) = 8 seconds.
+
+4)
+It will take log(N)/log(2) seconds. The plot is showed in the P5.png.
+
+5)
+It can be easily noticed that if there were 4 bags, it would take at least 3 seconds (including communication time) no matter how many cashiers there are. And when there were more than 4 bags, multiple cashiers would work faster than single cashier. Thus, we can know that for 256 bags we need 256/4=64 cashiers to count. Now we will have 64 number and then it will take log(64)/log(2)*2 = 12 seconds. Totally, it will take 3+12 = 15 seconds.
+
+6) 
+6a) To distribute all bags to each cashier, it will take at least log(256)/log(2)+1 = 9 seconds. And then there will be an extra second that cashiers need to tell the number to their partner (communicating). Now it turns to be the same question as 5). Thus, the total time should be 9+1+15=25 seconds.
```

Comment:
I'm not sure I follow your answer here, and I think it's a fair bit more complicated. This was how I did it:

We can write down a cost function for N bags and q workers as C(N, q) = q + (N/q - 1) + 2 * log_2(q). The first (q) term is the amount of time it takes to hand out the bags to the workers, it takes (N/q - 1) seconds for each worker to count their own bags, and then it takes 2 * log_2(q) to communicate and add the sums between the workers in parallel.

Following a prescription such as this, it's definitely not possible to get as low as 25 seconds.